internal/pkg/runtime/engine/apptainer/container_linux.go

```diff
@@ -628,6 +628,9 @@ func (c *container) mountGeneric(mnt *mount.Point, tag mount.AuthorizedTag) (err
 		dest = filepath.Join(c.session.FinalPath(), dest)
 	} else {
 		dest = mnt.Destination
+		if mnt.Type == "overlay" && c.session.GetReadonly() {
```

Comment:
This is detected by kernel overlay, all you need is to parse overlay mount options in the image driver and look if `upperdir` is present, if not you can consider this is a readonly mount